Add scripts to output contacts and stakeholders CSVs

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR adds two new scripts (`cyhy-contacts` and `cyhy-stakeholders`) that output CSV data for the contacts and stakeholders in the CyHy database.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

These scripts are very similar to those in [cisagov/ncats-webd](https://github.com/cisagov/ncats-webd) (see [here](https://github.com/cisagov/ncats-webd/blob/develop/ncats_webd/blueprints/stakeholders/contacts.py) and [here](https://github.com/cisagov/ncats-webd/blob/develop/ncats_webd/blueprints/stakeholders/stakeholders_info.py)), but those scripts were designed to be run as part of a web-based dashboard.  I modified them slightly so that they can be easily run from a shell.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I confirmed that these new scripts run as expected via a Docker container, which is the primary way that they are going to be used.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Notify the CyHy team that these new scripts are available for their use.

dav3r committed:
Add cyhy-contacts module

This code is mostly copied from ncats-webd, but modified to run as a standalone script.

dav3r committed:
Add cyhy-stakeholders module

This code is mostly copied from ncats-webd, but modified to run as a standalone script.

dav3r committed:
Add cyhy-contacts and cyhy-stakeholders to setup.py

dav3r committed:
Add cyhy-contacts and cyhy-stakeholders to docker-entrypoint.sh

dav3r committed:
Update README to show how to run cyhy-contacts and cyhy-stakeholders via Docker

dv4harr10 commented:
Hi Team, one medium severity issue at extras/create_snapshots_reports_scorecard.py line 815 and 819, the subprocess function has shell=True, which can allow a malicious actor to execute commands. Suggestion is to set shell=False instead.

jsf9k commented:
> Hi Team, one medium severity issue at extras/create_snapshots_reports_scorecard.py line 815 and 819, the subprocess function has shell=True, which can allow a malicious actor to execute commands. Suggestion is to set shell=False instead.

@dv4harr10 - I think [you already reported this issue](https://github.com/cisagov/cyhy-reports/issues/102).  In any event, it has nothing to do with the changes in this PR.

dv4harr10 commented:
Hi Team, one medium severity issue at extras/create_snapshots_reports_scorecard.py line 815 and 819, the subprocess function has shell=True, which can allow a malicious actor to execute commands. Suggestion is to set shell=False instead.

@dv4harr10 - I think https://github.com/cisagov/cyhy-reports/issues/102. In any event, it has nothing to do with the changes in this PR.

Ok Great, otherwise looks fine to me.

mcdonnnj commented on cyhy_report/contacts/contacts.py:
```diff
@@ -0,0 +1,65 @@
+    writer = DictWriter(output, fields)
+    writer.writeheader()
+
+    for doc in all_request_docs:
```
Is this doable? It just seems silly to assign the result to a variable and then iterate over it separately. I added a comment just to be as explanatory as the variable name.

```suggestion
    # Iterate through all request documents
    for doc in db.RequestDoc.find().sort("_id", 1):
```

mcdonnnj commented on cyhy_report/contacts/contacts.py:
```diff
@@ -0,0 +1,65 @@
+from cyhy.db import database
+
+def generate_contacts_csv(db):
+    all_request_docs = db.RequestDoc.find().sort("_id", 1)
```
If my previous suggestion is viable.

```suggestion
```

mcdonnnj commented on cyhy_report/contacts/contacts.py:
```diff
@@ -0,0 +1,65 @@
+                "Org ID": doc["_id"],
+                "Org Name": doc["agency"]["name"],
+                "Org Type": doc["agency"].get("type", "N/A"),
+                "Org Retired": doc.get("retired", False),
```
We could technically define this portion before the loop on contacts and do a `row.update()` for each contact but that might be an unnecessary micro-optimization.

dav3r replied:
Since you mentioned it, why not? 😄  Please see https://github.com/cisagov/cyhy-reports/pull/108/commits/befbd442d5570cd044f13273fa85fd8860c45fc4.

mcdonnnj commented on cyhy_report/stakeholders/stakeholders.py:
```diff
@@ -0,0 +1,129 @@
+
+        wr.writerow(
+            (
+                "{}".format(org["_id"]),
```
Is there a reason we're using `format()` for all of these when it's just doing string conversion at most?

dav3r replied:
Nope- I removed the `format()` jazz in https://github.com/cisagov/cyhy-reports/pull/108/commits/4317bc59c5310a350f30c10f59ae8bea6ed16378.

mcdonnnj commented on cyhy_report/stakeholders/stakeholders.py:
```diff
@@ -0,0 +1,129 @@
+            "First Scan",
+        )
+    )
+    stakeholders = db.RequestDoc.find({"_id": {"$in": stakeholder_ids}}).sort(
+        [("_id", 1)]
+    )
+    for org in stakeholders:
```
Same thought about seemingly unnecessary variable assignment.

```suggestion
    for org in db.RequestDoc.find({"_id": {"$in": stakeholder_ids}}).sort([("_id", 1)]):
```

mcdonnnj commented on cyhy_report/contacts/contacts.py:
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python
```
I know this is cribbed but we shouldn't be doing this anymore as the script in `bin/` negates the need.

```suggestion
```

mcdonnnj commented on cyhy_report/contacts/contacts.py:
```diff
@@ -0,0 +1,65 @@
+    args = docopt(__doc__, version="v0.0.1")
+    db = database.db_from_config(args["--section"])
+    print(generate_contacts_csv(db).getvalue())
+
+
+if __name__ == "__main__":
+    main()
```
I know this is cribbed but we shouldn't be doing this anymore as the script in `bin/` negates the need.

```suggestion
```

mcdonnnj commented on cyhy_report/stakeholders/stakeholders.py:
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python
```
I know this is cribbed but we shouldn't be doing this anymore as the script in `bin/` negates the need.

```suggestion
```

mcdonnnj commented on cyhy_report/stakeholders/stakeholders.py:
```diff
@@ -0,0 +1,129 @@
+    args = docopt(__doc__, version="v0.0.1")
+    db = database.db_from_config(args["--section"])
+    print(generate_stakeholders_csv(db).getvalue())
+
+
+if __name__ == "__main__":
+    main()
```
I know this is cribbed but we shouldn't be doing this anymore as the script in `bin/` negates the need.

```suggestion
```

mcdonnnj commented on cyhy_report/contacts/contacts.py:
I would prefer that this file is not executable as it should never be run directly with the other changes I've suggested.

dav3r replied:
I removed the executable bit in https://github.com/cisagov/cyhy-reports/pull/108/commits/75eea8b5f734d664fdb21e57b4ce12b75e32fef2.

mcdonnnj commented on cyhy_report/stakeholders/stakeholders.py:
I would prefer that this file is not executable as it should never be run directly with the other changes I've suggested.

dav3r replied:
I removed the executable bit in https://github.com/cisagov/cyhy-reports/pull/108/commits/75eea8b5f734d664fdb21e57b4ce12b75e32fef2.

mcdonnnj reviewed (changes requested):
Overall this looks like strong work 💪💪💪 I have some feedback items for your consideration.

dav3r committed:
Remove unnecessary code

This code is not needed since these functions are meant to be called via bin/cyhy-contacts and bin/cyhy-stakeholders.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove executable bit

The code in this files is meant to be executed via bin/cyhy-contacts and bin/cyhy-stakeholders

dav3r committed:
Remove two single-use variables

There's no need to create these variables- it's simpler to just iterate over the results of the respective database queries.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove unnecessary format() calls

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Slightly optimize CSV row creation loop

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj reviewed (changes requested):
Thank you for resolving my previous feedback. This is into nitpick turf but I'd really like it if we had docstrings for any functions in the new subpackages. This is consistent with any newer Python project of ours because of the inclusion of the [flake8-docstrings](https://github.com/pycqa/flake8-docstrings) plugin in our linting configuration. I know we cannot use our normal linting configuration due to the reliance on Python 2 for this project but I would like the code to be as close to passing our usual linter selection as reasonably possible.

dav3r committed:
Add pydoc strings to all functions in contacts.py and stakeholders.py

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r commented:
> Thank you for resolving my previous feedback. This is into nitpick turf but I'd really like it if we had docstrings for any functions in the new subpackages.

I added docstrings in https://github.com/cisagov/cyhy-reports/pull/108/commits/379a045a3e7ccc77d9632403e1723daf4ca3dd6c - please take a look and let me know if you have any other concerns.

mcdonnnj commented on cyhy_report/stakeholders/stakeholders.py:
```diff
@@ -0,0 +1,127 @@
+from cyhy.db import database
+
+def get_first_snapshot_times(db, owners):
+    """Return a dictionary with first snapshot times for a list of CyHy owner IDs
```
```suggestion
    """Return a dictionary with first snapshot times for a list of CyHy owner IDs.
```

mcdonnnj reviewed (approved):
Strong work 💪💪💪

dav3r committed:
Add a period to the end of a docstring sentence

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
